### PR TITLE
Research Station, HPCA and Computation System

### PR DESCRIPTION
**Pull request by serenibyss**
PR description will take hours to write so i will do it later. Opening for now (as a draft) since it is in a place where it can be tested in-game

### CODE REVIEW

**Review by ALongStringOfNumbers — COMMENTED** [review]
Looks like auto build fails with the research station, although not sure why, its seems like a simple enough structure for the auto build to complete

File: src/main/java/gregtech/api/recipes/recipeproperties/ComputationProperty.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    public static ComputationProperty getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new ComputationProperty();
+        }
+        return INSTANCE;
+    }
+
+    @Override
+    public void drawInfo(Minecraft minecraft, int x, int y, int color, Object value) {
+        minecraft.fontRenderer.drawString(I18n.format("gregtech.recipe.computation_per_tick", castValue(value)), x, y, color);
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/recipes/recipeproperties/ComputationProperty.java:25]: Maybe this should state if the CWUT/t required is Steady or Sporadic, if that is possible

File: src/main/java/gregtech/api/recipes/RecipeMaps.java [hunk 2]
```diff
@@ -1029,6 +1029,14 @@ public class RecipeMaps {
             .setProgressBar(GuiTextures.PROGRESS_BAR_REPLICATOR, MoveType.HORIZONTAL)
             .setSound(GTSoundEvents.REPLICATOR);
 
+    @ZenProperty
+    public static final RecipeMap<ComputationRecipeBuilder> RESEARCH_STATION_RECIPES = new RecipeMapResearchStation<>("research_station", 2, 1, 0, 0, new ComputationRecipeBuilder(), false)
+            .setProgressBar(GuiTextures.PROGRESS_BAR_ARROW, MoveType.HORIZONTAL)
+            .setSlotOverlay(false, false, GuiTextures.SCANNER_OVERLAY)
+            .setSlotOverlay(true, false, GuiTextures.RESEARCH_STATION_OVERLAY)
+            .setSound(GTSoundEvents.COMPUTATION);
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/recipes/RecipeMaps.java:1037]: I still say we should get science april fools

File: src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityDataBank.java [hunk 3]
```diff
@@ -195,6 +201,10 @@ public ICubeRenderer getBaseTexture(IMultiblockPart sourcePart) {
     @Override
     public void renderMetaTileEntity(CCRenderState renderState, Matrix4 translation, IVertexOperation[] pipeline) {
         super.renderMetaTileEntity(renderState, translation, pipeline);
+        renderTextures(renderState, translation, pipeline);
+    }
+
+    protected void renderTextures(CCRenderState renderState, Matrix4 translation, IVertexOperation[] pipeline) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityDataBank.java:207]: Why this change? Overriders could just override `renderMTE` as usual

File: src/main/java/gregtech/api/gui/widgets/BlockableSlotWidget.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            GlStateManager.colorMask(true, true, true, true);
+            GlStateManager.enableDepth();
+            GlStateManager.enableBlend();
+        }
+    }
+
+    @Override
+    public boolean isMouseOverElement(int mouseX, int mouseY) {
+        // prevent slot removal and hover highlighting when slot is blocked
+        return super.isMouseOverElement(mouseX, mouseY) && !isBlocked.get();
```
[the remark this answers is not in the document]
Comment (serenibyss) [reply src/main/java/gregtech/api/gui/widgets/BlockableSlotWidget.java:63]: I purposefully disabled the tooltip because I wanted to make it very clear the item could not be removed and give no indication of any interaction with the slot

File: src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityDataBank.java (hunk 3 above)
Comment (serenibyss) [reply src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityDataBank.java:207]: So that a subclass (Network Switch) could override this properly without also skipping on the super class's base texture rendering

File: src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityHPCA.java [hunk 5]
```diff
@@ -0,0 +1,748 @@
...
+        }
+    }
+
+    @Override
+    protected @NotNull BlockPattern createStructurePattern() {
+        return FactoryBlockPattern.start()
+                .aisle("AA", "CC", "CC", "CC", "AA")
+                .aisle("VA", "XV", "XV", "XV", "VA")
+                .aisle("VA", "XV", "XV", "XV", "VA")
+                .aisle("VA", "XV", "XV", "XV", "VA")
```
[the remark this answers is not in the document]
Comment (serenibyss) [reply src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityHPCA.java:169]: Honestly I prefer the original way just for a better visual indication of the structure in code

File: src/main/java/gregtech/api/recipes/recipeproperties/ComputationProperty.java (hunk 1 above)
Comment (serenibyss) [reply src/main/java/gregtech/api/recipes/recipeproperties/ComputationProperty.java:25]: I think it would take too many words to explain this meaningfully, so we should explain it elsewhere if anywhere

**Review by ALongStringOfNumbers — APPROVED** [review]
